emma-language/src/main/scala/org/emmalanguage/io/csv/CSVColumn.scala

```diff
@@ -0,0 +1,132 @@
+import scala.math.BigInt
+import scala.math.BigDecimal
+import scala.reflect.ClassTag
+
+/**
+ * Codec (provides encoding and decoding) type-class for CSV columns.
+ * A CSV column is an opaque value within a CSV record, encoded as a String.
+ * E.g. a CSV record of only one column would need no column delimiter.
+ *
+ * This codec is unsafe. Upon failure (mostly when decoding), an exception may be thrown.
+ * A very common exception type is [[NumberFormatException]].
+ *
+ * Usually CSV column codecs are provided for types that can be encoded as and decoded from a
+ * String, but cannot be broken down to a fixed number of fields. Otherwise [[CSVConverter]] is
+ * better suited. Instances provided by default are Scala and Java primitives, enumerations and
+ * [[Option]].
+ */
+trait CSVColumn[T] {
+
+  /**
+   * Decodes an instance of `T` from a String.
+   * Decoding is unsafe as it may throw an exception if parsing failed.
+   * A very common exception type is [[NumberFormatException]].
+   */
+  def from(column: String)(implicit csv: CSV): T
+
+  /**
+   * Encodes an instance of `T` as a String.
+   * In many cases delegating to `value.toString` is sufficient.
+   */
+  def to(value: T)(implicit csv: CSV): String
+}
+
+/** Factory methods and implicit instances of CSV column codecs. */
+object CSVColumn {
+
+  /** Summons an implicit codec for type `T` available in scope. */
+  def apply[T](implicit column: CSVColumn[T]): CSVColumn[T] = column
+
+  /** Creates a new codec for type `T` from pure conversion functions. */
+  def make[T](fromF: String => T, toF: T => String = (_: T).toString): CSVColumn[T] =
+    new CSVColumn[T] {
+      def from(column: String)(implicit csv: CSV) = fromF(column)
+      def to(value: T)(implicit csv: CSV) = toF(value)
+    }
+
+  /**
+   * Derives a new codec for type `B` from an existing codec for type `A`,
+   * given an isomorphism between `A` and `B`.
+   */
+  def iso[A, B](fromF: A => B, toF: B => A)(implicit A: CSVColumn[A]): CSVColumn[B] =
+    new CSVColumn[B] {
+      def from(column: String)(implicit csv: CSV) = fromF(A from column)
+      def to(value: B)(implicit csv: CSV) = A to toF(value)
+    }
+
+  // Scala primitives
+  implicit val stringCSVColumn:     CSVColumn[String]     = make(identity)
+  implicit val charCSVColumn:       CSVColumn[Char]       = make(_.head)
+  implicit val booleanCSVColumn:    CSVColumn[Boolean]    = make(_.toBoolean)
+  implicit val byteCSVColumn:       CSVColumn[Byte]       = make(_.toByte)
+  implicit val shortCSVColumn:      CSVColumn[Short]      = make(_.toShort)
+  implicit val intCSVColumn:        CSVColumn[Int]        = make(_.toInt)
+  implicit val longCSVColumn:       CSVColumn[Long]       = make(_.toLong)
+  implicit val floatCSVColumn:      CSVColumn[Float]      = make(_.toFloat)
+  implicit val doubleCSVColumn:     CSVColumn[Double]     = make(_.toDouble)
+  implicit val unitCSVColumn:       CSVColumn[Unit]       = make(const(()), const(""))
```

Comment:
The only way I could think of is to stratify the type classes as follows

```scala
trait CSVColumn // top level class
trait CSVSimple extends CSVColumn
trait CSVNullable extends CSVColumn
class CSVUnit extends CSVNullable
class CSVOption[T : CSVSimple] extends CSVNullable
```

I suggest to leave this as it is for now and rethink this in the future. I am also not sure whether it makes sense to provide codecs for collection types, as these are strictly speaking not part of the CSV spec.